src/v/bytes/iobuf.cc

```diff
@@ -102,6 +102,30 @@ bool iobuf::operator==(const iobuf& o) const {
     return true;
 }
 
+bool iobuf::operator<(const iobuf& o) const {
+    auto lhs = byte_iterator(cbegin(), cend());
+    auto lhs_end = byte_iterator(cend(), cend());
+    auto rhs = byte_iterator(o.cbegin(), o.cend());
+    auto rhs_end = byte_iterator(o.cend(), o.cend());
+    while (lhs != lhs_end && rhs != rhs_end) {
+        char l = *lhs;
+        char r = *rhs;
+        if (l < r) {
+            return true;
+        }
+        if (l > r) {
+            return false;
+        }
+        ++lhs;
+        ++rhs;
+    }
+    if (rhs != rhs_end) {
+        // lhs is a prefix of rhs.
+        return true;
+    }
+    return false;
```

Comment:
if you add `operator<` to `details::io_fragment` using `std::lexiographic_compare` then this code is simpler, but also there is probably more opportunity for the compiler to optimize down to `memcmp` or something efficient compared to what byte_iterator allows for?

Reply:
they are not the same size. but operator< would still be valid for the fragment. here in iobuf::operator< i think you'd have to do have some extra logic. maybe it's easier to operate directly on the fragments rather than trying to use fragment::operator<

Reply:
Yea i'm thinking about this case

buf1 [a b c]
buf2 [a] [z]

even if we had a fragment comparator, I'm not seeing how we could compare [z] against just [b c]

Reply:
For posterity, we chatted about this and what I was missing is that we wouldn't rely solely on the fragment comparator. Instead, the we'd iterate over fragments but keep track of indices within the iobuf/fragments that we could use `memcmp` to compare.

I'm going to punt on this for the purposes of unblocking other Iceberg work, but agree this is something worth addressing given iobuf is used widely.

FWIW we should probably do the same thing for operator==, which I modeled this method after.